File: solutions/c/queen-attack/1/queen_attack.c
#include "queen_attack.h"
#include <stdlib.h>

attack_status_t can_attack(position_t queen_1, position_t queen_2) {
    // Проверяем, что позиции находятся в пределах доски и не совпадают
    if ((queen_1.row > 7) || (queen_1.column > 7) ||
        (queen_2.row > 7) || (queen_2.column > 7) ||
        (queen_1.row == queen_2.row && queen_1.column == queen_2.column)) {
        return INVALID_POSITION;
    }

    // Проверка на одну линию (горизонталь или вертикаль)
    if (queen_1.row == queen_2.row || queen_1.column == queen_2.column) {
        return CAN_ATTACK;
    }

    // Проверка на одну диагональ
    if (abs(queen_1.row - queen_2.row) == abs(queen_1.column - queen_2.column)) {
        return CAN_ATTACK;
    }

    // Королевы не могут атаковать друг друга
    return CAN_NOT_ATTACK;
}

File: solutions/go/queen-attack/1/queen_attack.go
package queenattack

import (
    "math"
    "errors"
)

func CanQueenAttack(whitePosition, blackPosition string) (bool, error) {
	if len(whitePosition) != 2 || len(blackPosition) != 2 ||
    	(whitePosition[0] - 'a') > 7 || (whitePosition[1] - '1') > 7 ||
        (blackPosition[0] - 'a') > 7 || (blackPosition[1] - '1') > 7 ||
        (whitePosition[0] == blackPosition[0] && whitePosition[1] == blackPosition[1]) {
        return false, errors.New("INVALID_POSITION")
    }
    if whitePosition[0] == blackPosition[0] || whitePosition[1] == blackPosition[1] {
        return true, nil
    }
    if math.Abs(float64(int(whitePosition[0]) - int(blackPosition[0]))) == math.Abs(float64(int(whitePosition[1]) - int(blackPosition[1]))) {
        return true, nil
    }
    
    return false, nil
}
